fendermint/vm/topdown/src/sync.rs

```diff
@@ -0,0 +1,211 @@
+    let mut interval = tokio::time::interval(Duration::from_secs(config.polling_interval_secs));
+
+    loop {
+        interval.tick().await;
+
+        let starting_height = get_starting_height(provider).await?;
+        let latest_height = agent_proxy
+            .get_chain_head_height()
+            .await
+            .context("cannot fetch parent chain head")?;
+        if latest_height < config.chain_head_delay {
+            tracing::debug!("latest height not more than the chain head delay");
+            continue;
+        }
+        let ending_height = latest_height - config.chain_head_delay;
+
+        tracing::debug!("starting height: {starting_height}, ending_height: {ending_height}");
+
+        // we are going backwards in terms of block height, the latest block height is lower
+        // than our previously fetched head. It could be a chain reorg. We clear all the cache
+        // in `provider` and start from scratch
+        if starting_height > ending_height {
+            todo!()
+        }
+
+        let new_parent_views =
+            get_new_parent_views(agent_proxy, starting_height, ending_height).await?;
+        tracing::debug!("new parent views: {new_parent_views:?}");
+
+        atomically_or_err::<_, Error, _>(move || {
+            for (height, block_hash, validator_set, messages) in new_parent_views.clone() {
+                provider.new_parent_view(height, block_hash, validator_set, messages)?;
+            }
+            Ok(())
+        })
+        .await?;
+
+        tracing::debug!("updated new parent views till height: {ending_height}");
+    }
+}
+
+/// Obtain the starting block height to perform the parent view update
+async fn get_starting_height<T: ParentFinalityProvider + Send + Sync + 'static>(
+    provider: &Arc<T>,
+) -> anyhow::Result<BlockHeight> {
+    let starting_height = atomically_or_err::<_, Error, _>(|| {
+        // we are adding 1 to the height because we are fetching block by block, we also configured
+        // the sequential cache to use increment == 1.
+        Ok(if let Some(h) = provider.latest_height()? {
+            h + 1
+        } else {
+            let last_committed_finality = provider.last_committed_finality()?;
+            last_committed_finality.height + 1
+        })
+    })
+    .await?;
+
+    Ok(starting_height)
+}
+
+/// Obtain the new parent views for the input block height range
+async fn get_new_parent_views(
+    agent_proxy: &Arc<IPCAgentProxy>,
+    start_height: BlockHeight,
+    end_height: BlockHeight,
+) -> anyhow::Result<Vec<(BlockHeight, BlockHash, ValidatorSet, Vec<CrossMsg>)>> {
+    let mut block_height_to_update = vec![];
+    for h in start_height..=end_height {
+        let block_hash = agent_proxy
+            .get_block_hash(h)
+            .await
+            .context("cannot fetch block hash")?;
+        let validator_set = agent_proxy
+            .get_validator_set(h)
+            .await
+            .context("cannot fetch validator set")?;
```

Comment:
It would be worth creating an issue to optimise this so that if there's no change then it only records the configuration ID. Like a method that says `get_validator_set_if_changed(h, current_id)` and it returns `None` if it's the same as the `current_id`, so we don't store a copy of the entire power table at every height in memory if nothing happened. cc @adlrocha

Reply:
I think for all configuration id related issues, we should handle them altogether